sqlglot/jsonpath.py

```diff
@@ -35,145 +35,144 @@ class JSONPathTokenizer(Tokenizer):
-    i = 0
-
-    def _curr() -> t.Optional[TokenType]:
-        return tokens[i].token_type if i < size else None
-
-    def _prev() -> Token:
-        return tokens[i - 1]
-
-    def _advance() -> Token:
-        nonlocal i
-        i += 1
-        return _prev()
-
-    def _error(msg: str) -> str:
-        return f"{msg} at index {i}: {path}"
-
-    @t.overload
-    def _match(token_type: TokenType, raise_unmatched: Lit[True] = True) -> Token:
-        pass
-
-    @t.overload
-    def _match(token_type: TokenType, raise_unmatched: Lit[False] = False) -> t.Optional[Token]:
-        pass
-
-    def _match(token_type, raise_unmatched=False):
-        if _curr() == token_type:
-            return _advance()
-        if raise_unmatched:
-            raise ParseError(_error(f"Expected {token_type}"))
-        return None
-
-    def _parse_literal() -> t.Any:
-        token = _match(TokenType.STRING) or _match(TokenType.IDENTIFIER)
-        if token:
-            return token.text
-        if _match(TokenType.STAR):
-            return exp.JSONPathWildcard()
-        if _match(TokenType.PLACEHOLDER) or _match(TokenType.L_PAREN):
-            script = _prev().text == "("
-            start = i
-
-            while True:
-                if _match(TokenType.L_BRACKET):
-                    _parse_bracket()  # nested call which we can throw away
-                if _curr() in (TokenType.R_BRACKET, None):
-                    break
-                _advance()
-
-            expr_type = exp.JSONPathScript if script else exp.JSONPathFilter
-            return expr_type(this=path[tokens[start].start : tokens[i].end])
-
-        number = "-" if _match(TokenType.DASH) else ""
-
-        token = _match(TokenType.NUMBER)
-        if token:
-            number += token.text
-
-        if number:
-            return int(number)
-
-        return False
-
-    def _parse_slice() -> t.Any:
-        start = _parse_literal()
-        end = _parse_literal() if _match(TokenType.COLON) else None
-        step = _parse_literal() if _match(TokenType.COLON) else None
-
-        if end is None and step is None:
-            return start
-
-        return exp.JSONPathSlice(start=start, end=end, step=step)
-
-    def _parse_bracket() -> exp.JSONPathPart:
-        literal = _parse_slice()
-
-        if isinstance(literal, str) or literal is not False:
-            indexes = [literal]
-            while _match(TokenType.COMMA):
-                literal = _parse_slice()
-
-                if literal:
-                    indexes.append(literal)
-
-            if len(indexes) == 1:
-                if isinstance(literal, str):
-                    node: exp.JSONPathPart = exp.JSONPathKey(this=indexes[0])
-                elif isinstance(literal, exp.JSONPathPart) and isinstance(
-                    literal, (exp.JSONPathScript, exp.JSONPathFilter)
-                ):
-                    node = exp.JSONPathSelector(this=indexes[0])
+    def parse(self, path: str) -> exp.JSONPath:
```

Comment:
FYI - didn't notice this was moved within the tokenizer class at first, discussed in slack about moving to top-level and adding a `dialect` kwarg to get the correct tokenizer for each dialect.